### PR TITLE
A Utility pass to convert external modules to empty modules

### PR DESCRIPTION
**Pull request by darthscsi**
When linting, it's handy to have no external modules.  This pass converts external modules to normal modules with emptyish bodies.

### CODE REVIEW

**Comment by drom** [comment]
image

File: lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+void RTLGreyBoxerPass::runOnOperation() {
+  auto topModule = getOperation().getBody();
+  SmallVector<rtl::RTLModuleExternOp, 8> toErase;
+  OpBuilder builder(topModule->getTerminator());
+
+  for (auto &op : *topModule)
```
Comment (youngar) [inline lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp:34]: I think you can use `llvm::make_early_inc_range(*topModule)` and then erase the modules as you go.

File: include/circt/Dialect/SV/SVPasses.td [hunk 2]
```diff
@@ -26,4 +26,16 @@ def RTLCleanup : Pass<"rtl-cleanup", "rtl::RTLModuleOp"> {
   let constructor = "circt::sv::createRTLCleanupPass()";
 }
 
+def RTLGreyBoxer : Pass<"rtl-grey-boxer", "mlir::ModuleOp"> {
```
Comment (lattner) [inline include/circt/Dialect/SV/SVPasses.td:29]: I'd recommend a more explicit name for this like "-stub-module-prototypes".  It isn't obvious what "grey boxing".

File: lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
+//===- RTLCleanup.cpp - RTL Cleanup Pass ----------------------------------===//
```
Comment (lattner) [inline lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp:1]: Plz update first line comment

File: lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp (hunk 1 above)
Comment (lattner) [reply lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp:34]: +1

File: lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+} // end anonymous namespace
+
+void RTLGreyBoxerPass::runOnOperation() {
+  auto topModule = getOperation().getBody();
+  SmallVector<rtl::RTLModuleExternOp, 8> toErase;
+  OpBuilder builder(topModule->getTerminator());
+
+  for (auto &op : *topModule)
+    if (auto module = dyn_cast<rtl::RTLModuleExternOp>(op)) {
+      module.dump();
```
Comment (lattner) [inline lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp:36]: Plz remove debug dumps

File: lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      SmallVector<rtl::ModulePortInfo, 8> ports;
+      module.getPortInfo(ports);
+      auto nameAttr = builder.getStringAttr(module.getName());
```
Comment (lattner) [inline lib/Dialect/SV/Transforms/RTLGreyBoxer.cpp:39]: You can use `module.getNameAttr()` (and if it doesn't exist, plz add it)

**Review by lattner — APPROVED** [review]

**Pull request merged**